src/cookie.rs

```diff
@@ -90,14 +95,17 @@ pub(crate) fn extract_response_cookies<'a>(
         .map(|value| Cookie::parse(value))
 }
 
-/// A persistent cookie store that provides session support.
-#[derive(Default)]
-pub(crate) struct CookieStore(pub(crate) cookie_store::CookieStore);
-
-impl<'a> fmt::Debug for CookieStore {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        self.0.fmt(f)
-    }
+use std::any::Any;
+/// Actions for a persistent cookie store providing session supprt.
+pub trait CookieStore: fmt::Debug + Send + Sync {
```

Comment:
Here's my suggestions around the actual trait:

- I'd remove the downcasting stuff. Instead of asking the `Client` for a reference, I would suggest that implementors could provide a type that has an `Arc` of the data or something, and they check their own clone from time to time.
- We could move the lock stuff to the implementation, by making the store method also take `&self`. I say this just since if people already have an `Arc<Mutex>`, it seems wasteful to then wrap *that* in an `RwLock`.
- Method names `set_cookies` and `cookies()`?
- I wonder about using `Vec`s of strings, versus slices of `HeaderValue`s... I haven't thought this one through too far.
- Probably remove the `Debug` bounds, at least.